design-proposals/helm-chart.md

```diff
@@ -0,0 +1,249 @@
+
+```text
+├── Chart.yaml
+├── templates
+│    ├── kubevirt-crd.yaml
+│    └── kubevirt-operator.yaml
+└── values.yaml
+```
+
+```text
+├── Chart.yaml
+├── templates
+│   └── kubevirt-cr.yaml
+└── values.yaml
+```
+
+This approach is also similar to the current installation / uninstallation flows:
+
+```shell
+$ helm install kubevirt-operator <repo-url>
+$ helm install kubevirt-cr <repo-url>
+...
+$ helm uninstall kubevirt-cr
+$ sleep 45 # wait for the operator to uninstall the different components
+$ helm uninstall kubevirt-operator
+```
+
+Note that the CRD is under `/templates` for this example.
+This means that we can execute upgrades / deletes without custom hooks.
+This isn't possible if the CR is also a template in the same chart.
+
+### Approach N: Exotic options
+
+As seen above, Helm provides a lot of customization possibilities.
+I'd only mention some more and move on.
+Let me know if you're interested in either of the following:
+- One CRD + Operator chart, post-install hook creating the CR
+- One CRD, one CR and one Operator chart
+
+### Overview
+
+Many Helm chart layouts can work. Choosing one depends on how much we'd want to rely on chart hooks.
+
+[Approach 1](#approach-1-single-chart) was chosen for the initial Helm chart linked above in order to:
+- stay as close as possible to the original manifest setup
+- demonstrate chart hooks in action
+
+As mentioned earlier, I believe [Approach 2](#approach-2-one-crd-chart-and-one-cr--operator-chart) is the most popular so far.
+
+[Approach 3](#approach-3-one-cr-chart-and-one-crd--operator-chart) is most similar to the current setup.
+
+The CDI project has the same `CRD-Operator-CR` setup so all of the above applies to it as well.
+
+Note that the file structures above are purely for illustrative purposes.
+Different Kubernetes resources (e.g. `rbac`) could be extracted as separate templates (files) in order to keep the charts cleaner and easier to navigate.
+
+## Chart Management
+
+The second problem which needs to be addressed is how Helm charts are going to be created and modified.
+
+Currently, the manifests in KubeVirt are generated in two steps:
+1. `resource-generator` generates semi-populated configurations under `/manifests/generated`
+2. `manifest-templator` produces the final version of the manifests which are then shipped as part of each GitHub release
+
+### Approach 1: Generate Helm chart(s)
+
+Helm charts are in general not that different from plain Kubernetes manifests.
+It all depends on how much customization we'd allow the user but for the SUSE Edge initial chart we kept as little modifications as possible.
+
+We should be able to plug in the functionality to generate a `Chart.yaml`, `values.yaml` and possibly hooks and organize those in the necessary file structure.
+The charts can then be generated, packaged and pushed to the configured Helm repository at build time.
+
+### Approach 2: Manually curate a Helm chart
```

Comment:
Helm charts in fact use the content of the manifests. Since this content is generated for every build, it should be relatively easy to also structure it properly for Helm and produce the related artifacts. Those can be released as `*.tgz` or via `quay.io`. IMHO, it would be a better way, comparing to manual maintenance in a separate repository.

Reply:
The `Generate Helm chart(s)` definitely has the better end result. However, most charts are highly customisable and even if we don't take such approach, I imagine we'd still template some values so the manifests will not be completely the same.

This way the users would be able to work with their preferred configuration without relying on `kubectl patch` and / or manually modifying manifests before applying them e.g. configuring custom options for the CR would be possible via something like `$ helm install kubevirt-cr <repo-url> --set configuration.developerConfiguration.cpuAllocationRatio=5`

Reply:
I am not a kubevirt developer, but I had a thought.

I understand the motivation for generating one from the other.  It's [DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself), right?  But generating a helm chart from the manifests seems backwards to me.  Couldn't you generate manifests from the helm chart, instead, using `helm template` and default values?

Patching templates into manifests sounds like it will become tricky to keep up with changes to the manifest in future releases.  Working in the other direction seems trivial to me (just apply the template).

That way, you'd get:
* a single source of truth (the helm chart)
* an easy, low-maintenance release process (just run `helm template`)
* guaranteed consistency between the two installation methods
* consistency of template parameters and values across versions
    * no scramble to update patches for new versions
* ability to add as much customization as you want, without exponentially increasing complexity

I may be wrong, maybe it's a terrible idea.  But I was surprised to see no one mention it yet.